Add wrap for libccp4c

taw10 opened the pull request:
This adds a wrap for the C-language part of the [CCP4](https://www.ccp4.ac.uk/) core library.  The source code also includes a set of Fortran bindings, which are currently ignored but could be added as an optional extra if there were demand.

The license for this library is LGPL3, with liability disclaimer modified for UK law.  I've licensed my meson.build file as BSD to match the other wraps, but I'd also be happy to use LGPL3 instead with a view to eventually upstreaming the Meson file.  Note that LGPL applies only to this core library - the rest of the main CCP4 suite (which is huge), based on this library, has a proprietary licence.

The code is old (20 years, some parts older still) and there are a number of compiler warnings.  Following wrap guidelines, I haven't fixed them.  However, I did take the liberty of suppressing them with compiler options, to stop them from drowning warnings from the main project.

The only download source for the libraries is unfortunately an unreliable FTP server, so I've provided a more reliable fallback.  I've been in touch with the upstream authors to ask about hosting it on their [Github account](https://github.com/ccp4) and also fixing the compiler warnings.

jpakkane commented on subprojects/packagefiles/libccp4c/meson.build:
```diff
@@ -0,0 +1,104 @@
+
+
+cc = meson.get_compiler('c')
+mdep = cc.find_library('m', required: true)
```
This should be `required: false` to make it portable. Most platforms (in fact everything except Linux) have math functions in `libc` and don't have a standalone `m` library at all.

Also, `true` is the default value so this would be superfluous in any case.

taw10 replied:
Ok, fixed!

eli-schwartz commented on subprojects/packagefiles/libccp4c/meson.build:
```diff
@@ -0,0 +1,104 @@
+mdep = cc.find_library('m', required: true)
+
+
+# Note that the source code assumes that datadir='share'.
```
You can assert this I suppose... how does it handle different prefixes?

taw10 replied:
It literally does the following:
```
      env_file = ccp4_utils_joinfilenames(ccp4_utils_joinfilenames(
                  ccp4_utils_joinfilenames(pkgroot,"share"),"ccp4"),"environ.def");
```
Where `pkgroot` is set like this:
```
#ifdef PACKAGE_ROOT
  const char *pkgroot=PACKAGE_ROOT;
#else
  const char *pkgroot=get_parent_directory(argv[0]);
#endif
```
And `PACKAGE_ROOT` is set by meson.build.  However, this can be bypassed by setting an environment variable (`CINCL`).  This environment variable would be set when using the "full" CCP4 installation, but I wouldn't expect it to be the case when using the libraries via the wrap.

However, the environ.def (also default.def) file isn't part of these libraries, but rather part of the full CCP4 installation which is under a proprietary license.  It's just a list of default values for various program arguments, so it's a bit mean to not include it in the LGPL part, but that's the way it is.  Therefore, the question is kind of theoretical because the installed file will almost certainly never be seen when the wrap is used.  Adding an assertion seems like it would prevent the top-level project from customising the paths, without really adding any security because the project is almost certainly not using these files.  If it were, it would be part of the full CCP4 project and the environment variable would be set.

There is a related problem for some of the data files that are installed by the package, e.g. `syminfo.lib`.  The code will only be able to find them if an environment variable is set to the right location, even though the real location could easily be set by the build system (there is no particular reason to want to override the file).  This is something I _really_ want to fix via an upstream PR, but cannot yet do.

It's awful, I agree.  I don't see any way to fix it without changes upstream.

eli-schwartz replied:
Ah, okay -- thanks for explaining.

xclaesse commented on subprojects/packagefiles/libccp4c/meson.build:
```diff
@@ -0,0 +1,104 @@
+add_project_arguments('-Wno-stringop-overflow', language: 'c')
+add_project_arguments('-Wno-format-overflow', language: 'c')
+add_project_arguments('-Wno-misleading-indentation', language: 'c')
+add_project_arguments('-Wno-pointer-compare', language: 'c')
```
Better only add warning flags your compiler actually understands: `add_project_arguments(cc.get_supported_arguments([...]), language: 'c')`

taw10 replied:
Nice hint - fixed!

xclaesse commented on subprojects/packagefiles/libccp4c/meson.build:
```diff
@@ -0,0 +1,104 @@
+
+# Note that the source code assumes that datadir='share'.
+# If it is not, programs using the library will not be able
+# to find 'environ.def'.
```
You don't seems to install `environ.def`.

taw10 replied:
I've hopefully addressed this with my comment on the review by @eli-schwartz - see above.

eli-schwartz commented:
Please rebase on master -- we've added a Windows CI check.

taw10 commented:
> Please rebase on master -- we've added a Windows CI check.

Ok, rebased.  I don't have the ability to test on Windows here, so I'm expecting some trouble from the CI.  Let's see...

eli-schwartz commented:
Sorry -- merging another PR broke the CI so I rebased it myself.

eli-schwartz commented:
Alright, yeah, this works on Linux but fails on Windows:

```
D:\a\wrapdb\wrapdb\subprojects\libccp4-6.5.1\ccp4\ccp4_sysdep.h(244): fatal error C1189: #error:  "Can't determine machine number format"
```

taw10 commented:
Hmmm, this would probably be possible to fix, but only with upstream involvement.  Building CCP4 (the whole project, not just this library) on Windows is reported to work  with MinGW or Cygwin, but not VS (see http://devtools.fg.oisin.rc-harwell.ac.uk/#windows).

Is this a deal-breaker?  If so, I'll close the PR for now, see what I can do upstream, and come back later!

eli-schwartz commented:
This is not a dealbreaker, but I would recommend adding an explicit

```
if cc.get_id() == 'msvc'
    error('libccp4c cannot be compiled using MSVC')
endif
```

This will make it obvious to people that they shouldn't bother trying to figure out if they have done something wrong to make it not work, and that yes, it is actually a known and expected issue.

jpakkane commented:
You need to edit `ci_config.json` so that it does not try to build it on Windows.

eli-schwartz commented:
According to README.md, the correct way to do this is to mark the wrap as `linux_only`.

The undocumented method you used instead is commented in the json file as "mark the wrap for automatic failure on Windows as known-broken".

The nuance here is that `skip_*` is used solely by the mass regression tests, and indicates that the wrap is ***known to fail and needs fixing*** and is always run in CI as normal, but, it is skipped by the mass regression tests because we cannot test for regressions in things that are already broken.

The actual documented functionality, on the other hand, marks it as "not intended to compile on Windows because it wasn't authored upstream to allow for that possibility".

taw10 commented:
Got it, sorry for the noise!
